EKF/EKFGSF_yaw_estimator.cpp

```diff
@@ -0,0 +1,827 @@
+		rot312(2) = atan2f(-_ahrs_ekf_gsf[model_index].R(2, 0), _ahrs_ekf_gsf[model_index].R(2, 2));  // third rotation is about Y and is taken from AHRS
+
+		// Calculate the body to earth frame rotation matrix
+		const Dcmf R = taitBryan312ToRotMat(rot312);
+
+		// update the quaternion used by the AHRS prediction algorithm
+		_ahrs_ekf_gsf[model_index].quat = Quatf(R);
+
+	} else {
+		// using a 321 Tait-Bryan rotation to define yaw state
+		// take roll pitch yaw from AHRS prediction
+		Eulerf euler321(_ahrs_ekf_gsf[model_index].R);
+
+		// replace the yaw angle using the EKF state estimate
+		euler321(2) = _ekf_gsf[model_index].X[2];
+
+		// update the quaternion used by the AHRS prediction algorithm
+		_ahrs_ekf_gsf[model_index].quat = Quatf(euler321);
+
+	}
+}
+
+void Ekf::initialiseEKFGSF()
+{
+	memset(&X_GSF, 0, sizeof(X_GSF));
+	_ekf_gsf_vel_fuse_started = false;
+	_emergency_yaw_reset_time = 0;
+	_time_last_on_ground_us = 0;
+	_yaw_extreset_counter = 0;
+	_do_emergency_yaw_reset = false;
+	_ekf_gsf_yaw_variance = sq(M_PI_2_F);
+	memset(&_ekf_gsf, 0, sizeof(_ekf_gsf));
+	const float yaw_increment = 2.0f * M_PI_F / (float)N_MODELS_EKFGSF;
+	for (uint8_t model_index = 0; model_index < N_MODELS_EKFGSF; model_index++) {
+		// evenly space initial yaw estimates in the region between +-Pi
+		_ekf_gsf[model_index].X[2] = -M_PI_F + (0.5f * yaw_increment) + ((float)model_index * yaw_increment);
+
+		// All filter models start with the same weight
+		_ekf_gsf[model_index].W = 1.0f / (float)N_MODELS_EKFGSF;
+
+		if (_control_status.flags.gps) {
+			_ekf_gsf[model_index].X[0] = _gps_sample_delayed.vel(0);
+			_ekf_gsf[model_index].X[1] = _gps_sample_delayed.vel(1);
+			_ekf_gsf[model_index].P[0][0] = sq(_gps_sample_delayed.sacc);
+			_ekf_gsf[model_index].P[1][1] = _ekf_gsf[model_index].P[0][0];
+		} else {
+			// no GPS so assume we are in a semi-static condition and moving no more that 0.5 m/s
+			_ekf_gsf[model_index].P[0][0] = sq(0.5f);
+			_ekf_gsf[model_index].P[1][1] = sq(0.5f);
+		}
+
+		// use half yaw interval for yaw uncertainty
+		_ekf_gsf[model_index].P[2][2] = sq(0.5f * yaw_increment);
+	}
+}
+
+void Ekf::runEKFGSF()
+{
+	// Iniitialise states first time
+	_ahrs_accel = _imu_sample_delayed.delta_vel / fmaxf(_imu_sample_delayed.delta_vel_dt, FILTER_UPDATE_PERIOD_S / 4);
+	if (!_ahrs_ekf_gsf_tilt_aligned) {
+		// check for excessive acceleration to reduce likelihood of large inital roll/pitch errors
+		// due to vehicle movement
+		const float accel_norm_sq = _ahrs_accel.norm_squared();
+		const float upper_accel_limit = CONSTANTS_ONE_G * 1.1f;
+		const float lower_accel_limit = CONSTANTS_ONE_G * 0.9f;
+		const bool ok_to_align = ((accel_norm_sq > lower_accel_limit * lower_accel_limit &&
+			  accel_norm_sq < upper_accel_limit * upper_accel_limit));
+		if (ok_to_align) {
+			initialiseEKFGSF();
+			alignQuatEKFGSF();
+			_ahrs_ekf_gsf_tilt_aligned = true;
+		}
+		return;
```

Comment:
Agree that that is the preferred solution, however I had a go at doing that at the time of writing this code and ran into compiler errors that I didn't know how to resolve, so have structured it the way you see here so that the algorithm can be properly tested.

If someone wants to provide assistance with converting this to a separate class, then that would be good.